plugins/linux-pipewire/pipewire-virtualcam.c

```diff
@@ -0,0 +1,318 @@
+		    &stream_events, params, 2, obs_pwvc)) {
+		blog(LOG_WARNING, "[pipewire]: Failed to create stream");
+		goto error;
+	}
+	return true;
+
+error:
+	blog(LOG_WARNING, "Failed to start virtual camera");
+	return false;
+}
+
+static void virtualcam_stop(void *data, uint64_t ts)
+{
+	UNUSED_PARAMETER(ts);
+	blog(LOG_INFO, "Virtual camera stopped");
+	struct obs_pipewire_virtualcam_data *obs_pwvc =
+		(struct obs_pipewire_virtualcam_data *)data;
+
+	obs_output_end_data_capture(obs_pwvc->output);
+	// stop pipewire stuff
+	obs_pw_stop_loop(&obs_pwvc->pw_core);
+	obs_pw_destroy_stream(&obs_pwvc->pw_stream);
+
+	obs_pw_destroy_context(&obs_pwvc->pw_core);
+	obs_pw_destroy_loop(&obs_pwvc->pw_core);
+}
+
+static void virtual_video(void *data, struct video_data *frame)
+{
+	struct obs_pipewire_virtualcam_data *obs_pwvc =
+		(struct obs_pipewire_virtualcam_data *)data;
+
+	// check if we have a running pipewire stream
+	if (!obs_pwvc->pw_stream.pw_stream_state) {
+		// blog(LOG_INFO, "No node connected");
+		return;
+	}
+
+	blog(LOG_DEBUG, "exporting frame to pipewire");
+	// get buffer
+	struct pw_buffer *pw_buf;
+	if ((pw_buf = pw_stream_dequeue_buffer(obs_pwvc->pw_stream.stream)) ==
+	    NULL) {
+		blog(LOG_WARNING, "pipewire: out of buffers");
+		return;
+	}
+
+	struct spa_buffer *spa_buf = pw_buf->buffer;
+	struct spa_data *d = spa_buf->datas;
+
+	if (spa_buf->n_datas != obs_pwvc->formatinfo.planes) {
+		blog(LOG_INFO,
+		     "pipewire: number of buffer planes is different");
+		return;
+	}
+	// copy data
+	for (unsigned int i = 0; i < obs_pwvc->formatinfo.planes; i++) {
+		if (d[i].data == NULL) {
+			blog(LOG_WARNING, "pipewire: buffer not mapped");
+			continue;
+		}
+		memcpy(d[i].data, frame->data[i],
+		       obs_pwvc->formatinfo.sizes[i]);
+		d[i].mapoffset = 0;
+		d[i].maxsize = obs_pwvc->formatinfo.sizes[i];
+		d[i].flags = SPA_DATA_FLAG_READABLE;
+		d[i].type = SPA_DATA_MemPtr;
+		d[i].chunk->offset = 0;
+		d[i].chunk->stride = obs_pwvc->formatinfo.strides[i];
+		d[i].chunk->size = obs_pwvc->formatinfo.sizes[i];
+	}
+
+	// add metadata
+	struct spa_meta_header *h;
+	if ((h = spa_buffer_find_meta_data(spa_buf, SPA_META_Header,
+					   sizeof(*h)))) {
+		h->pts = -1;
+		h->flags = 0;
+		h->seq = obs_pwvc->pw_stream.seq++;
+		h->dts_offset = 0;
+	}
+
+	// return buffer
+
+	blog(LOG_DEBUG, "********************");
+	blog(LOG_DEBUG, "pipewire: fd %lu", d[0].fd);
+	blog(LOG_DEBUG, "pipewire: dataptr %p", d[0].data);
+	blog(LOG_DEBUG, "pipewire: size %d", d[0].maxsize);
+	blog(LOG_DEBUG, "pipewire: stride %d", d[0].chunk->stride);
+	blog(LOG_DEBUG, "pipewire: width %d", obs_pwvc->formatinfo.width);
+	blog(LOG_DEBUG, "pipewire: height %d", obs_pwvc->formatinfo.height);
+	blog(LOG_DEBUG, "********************");
+
+	pw_stream_queue_buffer(obs_pwvc->pw_stream.stream, pw_buf);
+}
+
+void virtual_cam_register_output(void)
+{
+	struct obs_output_info pipewire_virtualcam_info = {
+		.id = "virtualcam_output",
```

Comment:
I think an id like "pwcam_output" or "pw_cam_output" would be better.

I don't list "pipewire_output" (that I use in #5383 as draft) because it is too confusing.
Once the name is set I change it in my PR to match.